Fix longpress copy

marcoandonosie opened the pull request:
**Problem**
Users couldn’t copy or play back chat messages from MessageView.

**Change**
- Added a long-press contextMenu to text bubbles with:
- Copy → writes message text to system clipboard (UIKit/AppKit).
- Speak → uses **AVSpeechSynthesizer** via a **@MainActor Speech** helper.
- Added matching accessibility rotor actions (**accessibilityAction(named:)**).


**Why**
- Improves usability and accessibility with minimal surface-area changes, keeping behavior scoped to MessageView.


**Docs**
- Inline comments added to MessageView.


**Testing**
- Manual: long-press menu shows Copy/Speak for text messages; tool messages show neither.
- Verified clipboard contents and TTS playback on simulator/device.
- Accessibility: rotor exposes both actions.


**Issue**
Fixes #26 
<img width="365" height="467" alt="Screenshot 2025-08-14 at 3 01 42 AM" src="https://github.com/user-attachments/assets/0d145347-ce4e-4752-924e-8a50f02e66b7" />
(Good First Issue).

jdisho commented:
Thanks for putting in the effort here. Since I opened the issue, I will leave some quick feedback:
- Since you are already investing time to make this work, let’s also be more considerate about the UI. From your screenshot, it looks a bit off. We should aim for something closer to what I shared as a screenshot in the issue.
- This PR doesn’t fully resolve #26 yet, because we also need to support Selection. That is especially useful when the LLM response is long and we only want to copy part of it.

lukaskollmer commented on Sources/SpeziChat/MessageView.swift:
```diff
@@ -76,6 +93,41 @@ public struct MessageView: View {
+                                        NSPasteboard.general.setString(t, forType: .string)
+                                        #endif
+                                    }
+                                    Button(String(localized: "Speak")) {   // <- no #if around this button
```
we might want to add SFSymbols to the buttons? eg: have `document.on.document` for the copy action and `speaker.wave.3` or `waveform` for the speak action?

lukaskollmer commented on Sources/SpeziChat/MessageView.swift:
```diff
@@ -76,6 +93,41 @@ public struct MessageView: View {
+                            }
+                        //voiceover option for long press
+                        #if canImport(AVFoundation)
+                            .accessibilityAction(named: Text(String(localized: "Speak"))) {
```
same as in the other place, the `String(localized:)` is redundant in SwiftUI initializers such as `Text`, `Button`, `Toggle`, etc

lukaskollmer commented on Sources/SpeziChat/MessageView.swift:
```diff
@@ -96,6 +148,31 @@ public struct MessageView: View {
 
+//speech helper function for playback of text in long press
+#if canImport(AVFoundation)
+import AVFoundation
```
no need to import this again

lukaskollmer reviewed (changes requested):
thanks for tackling this issue @marcoandonosie!

i'm in favor of merging this, but there's a couple of things that should be adjusted:
- we already have a [SpeziSpeech](https://github.com/StanfordSpezi/SpeziSpeech) package, which might be preferable over directly using the `AVSpeechSynthesizer` in here
- i agree with @jdisho's comment about the UI. UIKit has [`UITargetedPreview`](https://developer.apple.com/documentation/uikit/uitargetedpreview) which lets one customise the preview that is displayed when interacting with a long-pressure-triggered context menu. i'm not saying that we should use this exact API here, but maybe SwiftUI has something similar?
- we can probably remove most of the inline comments?
- regarding @jdisho's point about the selection: IMO it might be fine to implement a general "copy everything" feature that doesn't let you select an exact range, and then add the selection-based copy feature in a separate PR. mainly bc there isn't really any standard way of doing this; i don't think adding the ability to do a region-based select directly in the bubble would be trivial? we might just want to add an extra "Select & Copy" option to the context menu and give it an explanatory SF Symbol (`selection.pin.in.out` or `character.cursor.ibeam`), and then have that open a sheet with a non-editable but selectable TextView displaying the chat message text that has "Cancel" and "Copy" buttons in its navigation bar?
- thanks for adding accessibility support!

let me know if you need support with anything

marcoandonosie commented:
Thanks for the feedback Lukas and Joan! I’ve added the requested UI icons, accessibility features, and copy/speak options, and cleaned up redundancies from the comments. Functionally, the copy and speech features should work. The one piece I couldn’t quite get looking right yet is the highlight UI on long press — I may circle back to refine that when I have some more time, but in the meantime I wanted to share this progress so it’s not blocking and others can build on it if helpful.
